// @vue/language-service/lib/plugins/vue-document-drop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const language_core_1 = require("@vue/language-core");
const shared_1 = require("@vue/shared");
const path = require("path-browserify");
const vue_extract_file_1 = require("../plugins/vue-extract-file");
const types_1 = require("../types");
function create(ts) {
    return {
        name: 'vue-document-drop',
        create(context) {
            let casing = types_1.TagNameCasing.Pascal; // TODO
            return {
                async provideDocumentDropEdits(document, _position, dataTransfer) {
                    if (document.languageId !== 'html')
                        return;
                    const [virtualCode, sourceFile] = context.documents.getVirtualCodeByUri(document.uri);
                    const vueVirtualCode = sourceFile?.generated?.code;
                    if (!virtualCode || !(vueVirtualCode instanceof language_core_1.VueGeneratedCode))
                        return;
                    let importUri;
                    for (const [mimeType, item] of dataTransfer) {
                        if (mimeType === 'text/uri-list') {
                            importUri = item.value;
                        }
                    }
                    if (!importUri?.endsWith('.vue'))
                        return;
                    let baseName = importUri.substring(importUri.lastIndexOf('/') + 1);
                    baseName = baseName.substring(0, baseName.lastIndexOf('.'));
                    const newName = (0, shared_1.capitalize)((0, shared_1.camelize)(baseName));
                    const { sfc } = vueVirtualCode;
                    const script = sfc.scriptSetup ?? sfc.script;
                    if (!script)
                        return;
                    const additionalEdit = {};
                    const code = [...(0, language_core_1.forEachEmbeddedCode)(vueVirtualCode)].find(code => code.id === (sfc.scriptSetup ? 'scriptSetupFormat' : 'scriptFormat'));
                    const lastImportNode = (0, vue_extract_file_1.getLastImportNode)(ts, script.ast);
                    let importPath = path.relative(path.dirname(document.uri), importUri)
                        || importUri.substring(importUri.lastIndexOf('/') + 1);
                    if (!importPath.startsWith('./') && !importPath.startsWith('../')) {
                        importPath = './' + importPath;
                    }
                    additionalEdit.changes ??= {};
                    additionalEdit.changes[context.documents.getVirtualCodeUri(sourceFile.id, code.id)] = [];
                    additionalEdit.changes[context.documents.getVirtualCodeUri(sourceFile.id, code.id)].push({
                        range: lastImportNode ? {
                            start: script.ast.getLineAndCharacterOfPosition(lastImportNode.end),
                            end: script.ast.getLineAndCharacterOfPosition(lastImportNode.end),
                        } : {
                            start: script.ast.getLineAndCharacterOfPosition(0),
                            end: script.ast.getLineAndCharacterOfPosition(0),
                        },
                        newText: `\nimport ${newName} from '${importPath}'`
                            + (lastImportNode ? '' : '\n'),
                    });
                    if (sfc.script) {
                        const edit = (0, vue_extract_file_1.createAddComponentToOptionEdit)(ts, sfc.script.ast, newName);
                        if (edit) {
                            additionalEdit.changes[context.documents.getVirtualCodeUri(sourceFile.id, code.id)].push({
                                range: {
                                    start: document.positionAt(edit.range.start),
                                    end: document.positionAt(edit.range.end),
                                },
                                newText: edit.newText,
                            });
                        }
                    }
                    return {
                        insertText: `<${casing === types_1.TagNameCasing.Kebab ? (0, shared_1.hyphenate)(newName) : newName}$0 />`,
                        insertTextFormat: 2,
                        additionalEdit,
                    };
                },
            };
        },
    };
}
exports.create = create;
