// @vue/language-service/lib/plugins/vue-codelens-references.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const language_core_1 = require("@vue/language-core");
function create() {
    return {
        name: 'vue-codelens-references',
        create(context) {
            return {
                provideReferencesCodeLensRanges(document) {
                    return worker(document.uri, virtualCode => {
                        const result = [];
                        for (const map of context.documents.getMaps(virtualCode) ?? []) {
                            for (const mapping of map.map.mappings) {
                                if (!mapping.data.__referencesCodeLens)
                                    continue;
                                result.push({
                                    start: document.positionAt(mapping.generatedOffsets[0]),
                                    end: document.positionAt(mapping.generatedOffsets[mapping.generatedOffsets.length - 1]
                                        + mapping.lengths[mapping.lengths.length - 1]),
                                });
                            }
                        }
                        return result;
                    });
                },
            };
            function worker(uri, callback) {
                const [virtualCode, sourceFile] = context.documents.getVirtualCodeByUri(uri);
                if (!(sourceFile?.generated?.code instanceof language_core_1.VueGeneratedCode) || !sourceFile)
                    return;
                return callback(virtualCode, sourceFile);
            }
        },
    };
}
exports.create = create;
